src/mavlink-router/endpoint.cpp

```diff
@@ -771,22 +796,45 @@ ssize_t UdpEndpoint::_read_msg(uint8_t *buf, size_t len)
 
 int UdpEndpoint::write_msg(const struct buffer *pbuf)
 {
+    if (tx_buf.len + pbuf->len > TX_BUF_MAX_SIZE)
+    {
+        log_debug("Dropping message, tx buffer full");
+        return 0;
+    }
+    memcpy(&tx_buf.data[tx_buf.len], pbuf->data, pbuf->len);
+    tx_buf.len += pbuf->len;
+
+    if (tx_buf.len >= _nagle_buf_max_packet_size)
+    {
+        int bytes_remaining = _write_internal_buffer();
+        if (bytes_remaining > 0)
+        {
+            _reschedule_write();
+        }
+    } else if (!_write_scheduled) {
+        _reschedule_write();
```

Comment:
the check for "_write_scheduled" and only conditionally calling "_reschedule_write" if false is not consistent

would suggest moving check into method -- so after this is called, write is scheduled

the most logic thing would be to have the deadline as parameter to the schedule call -- that way you can have different deadline based on packet class (see lorenz' comments)
the schedule function would then compare with "presently" set deadline from previous packets and reschedule to pick earliest ones
may want to support parameterizing this on maximum delay

Reply:
I'd prefer to work with deadlines too, but that would require more drastic mavlink-router surgery. I think we can work reasonably with the delay-based API just by doing a force_write() in the case where a high-priority message comes through. Yes, we lose the coalescing capabilities for that message type, but I generally see two classes of messages: general telemetry streaming which isn't sensitive to an extra 20ms of latency, and command-ack pairs which are sensitive to latency. I can't think of a use-case for in between.

Reply:
see other comment, if you cannot cancel relative timers, then you need to track deadline anyways to detect premature wakeups

that means you would need to track "absolute send deadline" in the endpoint class, while communicating "relative timeout" to the mainloop class without requiring major surgery there

I don't think this changes fundamentally when deadline is always either "immediate" or "now + 20ms"